tracing-appender/src/non_blocking.rs

```diff
@@ -0,0 +1,146 @@
+use crate::worker::Worker;
+use crossbeam_channel::{bounded, Sender};
+use std::io;
+use std::io::Write;
+use std::sync::atomic::Ordering;
+use std::sync::atomic::{AtomicBool, AtomicU64};
+use std::sync::Arc;
+use std::thread::JoinHandle;
+use tracing_subscriber::fmt::MakeWriter;
+
+pub const DEFAULT_BUFFERED_LINES_LIMIT: usize = 128_000;
+
+#[derive(Debug)]
+pub struct WorkerGuard {
+    guard: Option<JoinHandle<()>>,
+    shutdown_signal: Arc<AtomicBool>,
+}
+
+#[derive(Clone, Debug)]
+pub struct NonBlocking {
+    error_counter: Arc<AtomicU64>,
+    channel: Sender<Vec<u8>>,
+    is_lossy: bool,
+}
+
+impl NonBlocking {
+    pub fn new<T: Write + Send + Sync + 'static>(writer: T) -> (NonBlocking, WorkerGuard) {
+        NonBlockingBuilder::default().build(writer)
+    }
+
+    fn create<T: Write + Send + Sync + 'static>(
+        writer: T,
+        buffered_lines_limit: usize,
+        is_lossy: bool,
+    ) -> (NonBlocking, WorkerGuard) {
+        let (sender, receiver) = bounded(buffered_lines_limit);
+        let shutdown_signal = Arc::new(AtomicBool::new(false));
+
+        let worker = Worker::new(receiver, writer, shutdown_signal.clone());
+        let worker_guard = WorkerGuard::new(worker.worker_thread(), shutdown_signal);
+
+        (
+            Self {
+                channel: sender,
+                error_counter: Arc::new(AtomicU64::new(0)),
+                is_lossy,
+            },
+            worker_guard,
+        )
+    }
+
+    pub fn error_counter(&self) -> Arc<AtomicU64> {
+        self.error_counter.clone()
+    }
+}
+
+#[derive(Debug)]
+pub struct NonBlockingBuilder {
+    buffered_lines_limit: usize,
+    is_lossy: bool,
+}
+
+impl NonBlockingBuilder {
+    pub fn buffered_lines_limit(mut self, buffered_lines_limit: usize) -> NonBlockingBuilder {
+        self.buffered_lines_limit = buffered_lines_limit;
+        self
+    }
+
+    pub fn lossy(mut self, is_lossy: bool) -> NonBlockingBuilder {
+        self.is_lossy = is_lossy;
+        self
+    }
+
+    pub fn build<T: Write + Send + Sync + 'static>(self, writer: T) -> (NonBlocking, WorkerGuard) {
```

Comment:
Nit: other builders in `tracing` use `finish` rather than `build`...we should try to make this consistent.